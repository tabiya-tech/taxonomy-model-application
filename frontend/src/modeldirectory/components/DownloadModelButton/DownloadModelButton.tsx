import { IconButton, Typography } from "@mui/material";
import FileDownloadIcon from "@mui/icons-material/FileDownload";
import * as React from "react";

const uniqueId = "1623572c-456c-419e-9201-3c4971406fe1";
export const DATA_TEST_ID = {
  DOWNLOAD_MODEL_BUTTON: `download-model-button-${uniqueId}`,
};

export interface DownloadModelButtonProps {
  downloadUrl: string;
}

export function DownloadModelButton(props: Readonly<DownloadModelButtonProps>) {
  return (
    <IconButton
      href={props.downloadUrl}
      download={props.downloadUrl}
      sx={{
        margin: 0,
        paddingY: (theme) => theme.tabiyaSpacing.xs,
        paddingX: (theme) => theme.tabiyaSpacing.xs,
        borderRadius: (theme) => theme.tabiyaRounding.sm,
      }}
      color={"primary"}
      title={"Download"}
      data-testid={DATA_TEST_ID.DOWNLOAD_MODEL_BUTTON}
    >
      <FileDownloadIcon sx={{ lineHeight: 0 }} />
      <Typography color={"primary"} variant="caption">
        Download
      </Typography>
    </IconButton>
  );
}
export default DownloadModelButton;
