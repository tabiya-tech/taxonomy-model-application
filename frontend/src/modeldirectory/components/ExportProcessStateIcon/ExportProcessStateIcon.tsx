import ExportProcessStateAPISpecs from "api-specifications/exportProcessState";
import CloudDownloadIcon from "@mui/icons-material/CloudDownload";

import * as React from "react";
import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { PulsatingIcon } from "src/theme/PulsatingIcon/PulsatingIcon";

const uniqueId = "bae86ed9-33bf-4492-a0e2-f9c8bd112bae";
export const DATA_TEST_ID = {
  ICON_STATUS_PENDING: `icon-status-pending-${uniqueId}`,
  ICON_STATUS_COMPLETED: `icon-status-success-${uniqueId}`,
  ICON_STATUS_RUNNING: `icon-status-running-${uniqueId}`,
  ICON_STATUS_UNKNOWN: `icon-status-unknown-${uniqueId}`,
};

export type ExportStatusIconProps = {
  exportProcessState: ModelInfoTypes.ExportProcessState;
};

export function ExportProcessStateIcon(props: Readonly<ExportStatusIconProps>) {
  switch (props?.exportProcessState?.status) {
    case ExportProcessStateAPISpecs.Enums.Status.RUNNING:
      return (
        <PulsatingIcon
          icon={CloudDownloadIcon}
          titleAccess="Export in progress"
          color="info"
          data-testid={DATA_TEST_ID.ICON_STATUS_RUNNING}
        />
      );
    case ExportProcessStateAPISpecs.Enums.Status.PENDING:
      return (
        <CloudDownloadIcon titleAccess="Pending" color="disabled" data-testid={DATA_TEST_ID.ICON_STATUS_PENDING} />
      );
    case ExportProcessStateAPISpecs.Enums.Status.COMPLETED: {
      const result = props.exportProcessState.result;
      let title = "Export was successful";
      let color: "success" | "error" | "warning" = "success";

      if (result.exportWarnings) {
        color = "warning";
        title = "Export completed with warnings";
      }

      if (result.exportErrors) {
        color = "error";
        title = "Exported completed with errors";
      }

      if (result.errored) {
        color = "error";
        title = "Export failed";
      }

      return <CloudDownloadIcon titleAccess={title} data-testid={DATA_TEST_ID.ICON_STATUS_COMPLETED} color={color} />;
    }
    default:
      return <div title="Unknown export status" data-testid={DATA_TEST_ID.ICON_STATUS_UNKNOWN} />;
  }
}

export default ExportProcessStateIcon;
