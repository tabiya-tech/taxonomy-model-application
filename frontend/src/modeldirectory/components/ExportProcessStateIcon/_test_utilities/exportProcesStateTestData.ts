import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ExportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllExportProcessStatePermutations = (): ModelInfoTypes.ExportProcessState[] => {
  // generate all permutations of ExportProcessState status, errored, exportErrors, exportWarnings
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ExportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((exportErrors) => {
        allBooleans.forEach((exportWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              exportErrors: exportErrors,
              exportWarnings: exportWarnings,
            },
            downloadUrl: "https://example.com",
            timestamp: new Date(),
          });
        });
      });
    });
  });
  return allPermutations;
};
