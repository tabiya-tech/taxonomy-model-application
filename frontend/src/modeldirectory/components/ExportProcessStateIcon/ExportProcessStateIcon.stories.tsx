import type { Meta, StoryObj } from "@storybook/react";
import ExportProcessStateIcon from "./ExportProcessStateIcon";
import { getAllExportProcessStatePermutations } from "./_test_utilities/exportProcesStateTestData";
import React from "react";

const meta: Meta<typeof ExportProcessStateIcon> = {
  title: "ModelDirectory/ExportProcessStateIcon",
  component: ExportProcessStateIcon,
  tags: ["autodocs"],
  argTypes: {},
};

export default meta;
type Story = StoryObj<typeof ExportProcessStateIcon>;

export const Shown: Story = {
  render: () => {
    return <ShowAllExportProcessStatePermutations />;
  },
};

function ShowAllExportProcessStatePermutations() {
  return (
    <>
      {getAllExportProcessStatePermutations().map((exportProcessState) => {
        return (
          <p key={exportProcessState.id}>
            <ExportProcessStateIcon exportProcessState={exportProcessState} />
            <span>
              {(() => {
                // @ts-ignore
                delete exportProcessState.id;
                return JSON.stringify(exportProcessState);
              })()}
            </span>
          </p>
        );
      })}
      <p>
        {
          // @ts-ignore
          <ExportProcessStateIcon exportProcessState={undefined} />
        }
        <span>{"undefined"}</span>
      </p>
    </>
  );
}
