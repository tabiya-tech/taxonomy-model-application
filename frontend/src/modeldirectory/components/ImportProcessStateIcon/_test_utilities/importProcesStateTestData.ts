import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import { getMockId } from "src/_test_utilities/mockMongoId";

export const getAllImportProcessStatePermutations = (): ModelInfoTypes.ImportProcessState[] => {
  // generate all permutations of ImportProcessState status, errored, parsingErrors, parsingWarnings
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  const allBooleans = [true, false];

  const allPermutations: ModelInfoTypes.ImportProcessState[] = [];
  allStatuses.forEach((status) => {
    allBooleans.forEach((errored) => {
      allBooleans.forEach((parsingErrors) => {
        allBooleans.forEach((parsingWarnings) => {
          allPermutations.push({
            id: getMockId(allPermutations.length + 1),
            status,
            result: {
              errored,
              parsingErrors: parsingErrors,
              parsingWarnings: parsingWarnings,
            },
            createdAt: new Date(allPermutations.length),
            updatedAt: new Date(allPermutations.length),
          });
        });
      });
    });
  });
  return allPermutations;
};
