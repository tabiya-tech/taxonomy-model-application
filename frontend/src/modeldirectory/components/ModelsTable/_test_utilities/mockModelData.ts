import ModelInfoAPISpecs from "api-specifications/modelInfo";
import LocaleAPISpecs from "api-specifications/locale";
import ImportProcessStateAPISpecs from "api-specifications/importProcessState";
import ExportProcessStateAPISpecs from "api-specifications/exportProcessState";
import { ImportProcessStateEnums } from "api-specifications/importProcessState/enums";

import { v4 as uuidv4 } from "uuid";

import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { getMockId } from "src/_test_utilities/mockMongoId";
import { getRandomLorem, getRandomString, getTestString } from "src/_test_utilities/specialCharacters";
import { faker } from "@faker-js/faker";
import { CELL_MAX_LENGTH } from "src/modeldirectory/components/ModelsTable/ModelsTable";

export function getOneFakeModel(id?: number): ModelInfoTypes.ModelInfo {
  const model = getArrayOfFakeModels(1)[0];
  model.id = getMockId(id ?? 1);
  return model;
}

export function getArrayOfFakeModels(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: faker.location.country().substring(0, ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: faker.location
          .countryCode("alpha-3")
          .substring(0, LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description:
        i % 2 === 0
          ? getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH)
          : getRandomLorem(CELL_MAX_LENGTH / 2), // 50% chance of long description
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: faker.lorem.text().substring(0, ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: faker.system.semver().substring(0, ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getArrayOfFakeModelsMaxLength(count: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < count; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: `${i + 1}/${count} - ${getRandomLorem(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH)}`.substring(
        0,
        ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH
      ),
      locale: {
        UUID: uuidv4(),
        name: getRandomLorem(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getRandomLorem(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getRandomLorem(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0, // 50% chance of released
      releaseNotes: getRandomLorem(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getRandomLorem(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export function getOneRandomModelMaxLength(): ModelInfoTypes.ModelInfo {
  return getArrayOfRandomModelsMaxLength(1)[0];
}

export function getArrayOfRandomModelsMaxLength(number: number): ModelInfoTypes.ModelInfo[] {
  const models: ModelInfoTypes.ModelInfo[] = [];
  for (let i = 0; i < number; i++) {
    const model: ModelInfoTypes.ModelInfo = {
      id: getMockId(i),
      UUID: uuidv4(),
      UUIDHistory: [uuidv4()],
      name: getRandomString(ModelInfoAPISpecs.Constants.NAME_MAX_LENGTH),
      locale: {
        UUID: uuidv4(),
        name: getRandomString(LocaleAPISpecs.Constants.NAME_MAX_LENGTH),
        shortCode: getTestString(LocaleAPISpecs.Constants.LOCALE_SHORTCODE_MAX_LENGTH),
      },
      description: getTestString(ModelInfoAPISpecs.Constants.DESCRIPTION_MAX_LENGTH),
      released: i % 2 === 0,
      releaseNotes: getTestString(ModelInfoAPISpecs.Constants.RELEASE_NOTES_MAX_LENGTH),
      version: getTestString(ModelInfoAPISpecs.Constants.VERSION_MAX_LENGTH),
      createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
      updatedAt: new Date(),
      path: faker.internet.url(),
      tabiyaPath: faker.internet.url(),
      exportProcessState: [getOneFakeExportProcessState(i)],
      importProcessState: getOneFakeImportProcessState(i)
    };
    models.push(model);
  }
  return models;
}

export const fakeModel: ModelInfoTypes.ModelInfo = {
  UUID: "8d914eab-6f7d-4183-accc-a09f99887b39",
  UUIDHistory: ["8d914eab-6f7d-4183-accc-a09f99887b39"],
  description: "aw j zt   agrkasl dy ogtimpsauwumu l utrovthao syertm beawpxluhyudgzbbm",
  id: "000000000000000000000001",
  importProcessState: {
    id: "000000000000000000000001",
    result: {
      errored: true,
      parsingErrors: true,
      parsingWarnings: true,
    },
    status: ImportProcessStateEnums.Status.PENDING,
    createdAt: new Date("2023-10-18T17:35:10.571Z"),
    updatedAt: new Date("2023-10-18T17:35:12.571Z"),
  },
  exportProcessState: [
    {
      downloadUrl: "https://www.example.com/",
      id: "000000000000000000000001",
      result: {
        errored: true,
        exportErrors: true,
        exportWarnings: true,
      },
      status: ExportProcessStateAPISpecs.Enums.Status.PENDING,
      timestamp: new Date("2023-10-18T17:35:10.571Z"),
      createdAt: new Date("2023-10-18T17:35:10.571Z"),
      updatedAt: new Date("2023-10-18T17:35:12.571Z"),
    },
  ],
  locale: {
    UUID: "7c4ea00b-fbea-4a31-9c04-009c813f5f34",
    name: "Palestine",
    shortCode: "ETH",
  },
  name: "1/1 - fysmros tcgcnjbbrrev",
  path: "https://unwelcome-editorial.net/",
  releaseNotes: "Perferendis modi impedit necessitatibus a",
  released: true,
  tabiyaPath: "https://silver-vulture.biz",
  createdAt: new Date("2023-10-18T17:35:10.571Z"),
  updatedAt: new Date("2022-10-18T17:35:10.571Z"),
  version: "0.2.2",
};

export function getRandomImportStatus(id: number) {
  const allStatuses = Object.values(ImportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getRandomExportStatus(id: number) {
  const allStatuses = Object.values(ExportProcessStateAPISpecs.Enums.Status); // Assuming it's an enum with string values
  return allStatuses[id % allStatuses.length];
}

export function getOneFakeExportProcessState(i: number): ModelInfoTypes.ExportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomExportStatus(i),
    result: {
      errored: i % 2 === 0,
      exportErrors: i % 2 === 0,
      exportWarnings: i % 2 === 0,
    },
    downloadUrl: faker.internet.url(),
    timestamp: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}

export function getOneFakeImportProcessState(i: number): ModelInfoTypes.ImportProcessState {
  return {
    id: getMockId(10000 + i),
    status: getRandomImportStatus(i),
    result: {
      errored: i % 2 === 0,
      parsingErrors: i % 2 === 0,
      parsingWarnings: i % 2 === 0,
    },
    createdAt: new Date(Date.now() - i * 1000 * 60 * 60 * 24),
    updatedAt: new Date(),
  };
}
