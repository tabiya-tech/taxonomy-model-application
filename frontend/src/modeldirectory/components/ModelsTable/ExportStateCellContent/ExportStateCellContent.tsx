import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import { DownloadModelButton } from "src/modeldirectory/components/DownloadModelButton/DownloadModelButton";
import ExportProcessStateIcon from "src/modeldirectory/components/ExportProcessStateIcon/ExportProcessStateIcon";
import ExportProcessStateAPISpecs from "api-specifications/exportProcessState";
import * as React from "react";

function isSuccessfulExport(exportProcessState: ModelInfoTypes.ExportProcessState): boolean {
  return (
    exportProcessState &&
    exportProcessState.status === ExportProcessStateAPISpecs.Enums.Status.COMPLETED &&
    !exportProcessState.result.errored &&
    !exportProcessState.result.exportErrors &&
    !exportProcessState.result.exportWarnings
  );
}

interface ExportStateCellProps {
  model: ModelInfoTypes.ModelInfo;
}

const uniqueId = "fe7db48f-9846-478f-b3e0-b299fb30775e";
export const DATA_TEST_ID = {
  EMPTY_DIV: `empty-div-${uniqueId}`,
};

export function ExportStateCellContent(props: Readonly<ExportStateCellProps>) {
  // Can download only if the last export was successful and there were no errors or warnings
  const sortedExportProcessStates = props.model.exportProcessState.toSorted(
    (a, b) => a.timestamp.getTime() - b.timestamp.getTime()
  );

  if (sortedExportProcessStates.length === 0) {
    return <div data-testid={DATA_TEST_ID.EMPTY_DIV} />;
  }

  const latestExportProcessState = sortedExportProcessStates[sortedExportProcessStates.length - 1];

  if (isSuccessfulExport(latestExportProcessState)) {
    return <DownloadModelButton downloadUrl={latestExportProcessState.downloadUrl} />;
  }
  return <ExportProcessStateIcon exportProcessState={latestExportProcessState} />;
}

export default ExportStateCellContent;
