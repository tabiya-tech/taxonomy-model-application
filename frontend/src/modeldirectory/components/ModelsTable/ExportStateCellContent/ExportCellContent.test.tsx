// mock the console
import "src/_test_utilities/consoleMock";
import { render, screen } from "src/_test_utilities/test-utils";
import { DATA_TEST_ID, ExportStateCellContent } from "./ExportStateCellContent";
import DownloadModelButton from "src/modeldirectory/components/DownloadModelButton/DownloadModelButton";
import { getOneFakeModel } from "src/modeldirectory/components/ModelsTable/_test_utilities/mockModelData";
import ExportProcessStateAPISpecs from "api-specifications/exportProcessState";
import { getAllExportProcessStatePermutations } from "src/modeldirectory/components/ExportProcessStateIcon/_test_utilities/exportProcesStateTestData";
import ExportProcessStateIcon from "src/modeldirectory/components/ExportProcessStateIcon/ExportProcessStateIcon";
import { randomUUID } from "crypto";
import { ModelInfoTypes } from "src/modelInfo/modelInfoTypes";
import ExportProcessState = ModelInfoTypes.ExportProcessState;

// mock the ExportProcessStateIcon component
jest.mock("src/modeldirectory/components/ExportProcessStateIcon/ExportProcessStateIcon", () => {
  const mock = jest.fn(() => {
    return <div data-testid={"mock-ExportProcessState-icon"} />;
  });
  return {
    __esModule: true,
    ExportProcessStateIcon: mock,
    default: mock,
  };
});

jest.mock("src/modeldirectory/components/DownloadModelButton/DownloadModelButton", () => {
  const mock = jest.fn(() => {
    return <div data-testid={"mock-DownLoadButton-icon"} />;
  });
  return {
    __esModule: true,
    DownloadModelButton: mock,
    default: mock,
  };
});

function getAllNotSuccessfulExportProcessStates() {
  return getAllExportProcessStatePermutations().filter((exportProcessState) => {
    return (
      exportProcessState.status !== ExportProcessStateAPISpecs.Enums.Status.COMPLETED ||
      exportProcessState.result.errored ||
      exportProcessState.result.exportErrors ||
      exportProcessState.result.exportWarnings
    );
  });
}

describe("ExportStateCellContent", () => {
  beforeEach(() => {
    (console.error as jest.Mock).mockClear();
    (console.warn as jest.Mock).mockClear();
  });
  // when there is no export process state
  // THEN expect nothing to be rendered
  test("should render nothing when there is no export process state", () => {
    // GIVEN some model with no export process state
    const givenModel = getOneFakeModel(1);
    givenModel.exportProcessState = [];

    // WHEN the component is rendered
    render(<ExportStateCellContent model={givenModel} />);

    // THEN expect no errors or warning to have occurred
    expect(console.error).not.toHaveBeenCalled();
    expect(console.warn).not.toHaveBeenCalled();

    // AND expect and empty div to be rendered
    const actualElement = screen.getByTestId(DATA_TEST_ID.EMPTY_DIV);
    expect(actualElement).toBeInTheDocument();
    // AND to have no children
    expect(actualElement).toBeEmptyDOMElement();
  });

  describe("should render the download button when the last export was successful", () => {
    function getSuccessfulExportProcessState(timestamp: number) {
      return {
        id: randomUUID(),
        status: ExportProcessStateAPISpecs.Enums.Status.COMPLETED,
        result: {
          errored: false,
          exportErrors: false,
          exportWarnings: false,
        },
        downloadUrl: randomUUID(),
        timestamp: new Date(timestamp),
        createdAt: new Date(timestamp),
        updatedAt: new Date(new Date(timestamp).getTime() + 1000)
      };
    }

    const NumberOfNotSuccessfulExportProcessStates = getAllNotSuccessfulExportProcessStates().length;

    function getTestExportProcessStates(indexOfSuccessFull: number): ExportProcessState[] {
      const allStates = getAllNotSuccessfulExportProcessStates();
      // put the successful one in the middle of the list
      const l = allStates.slice(0, indexOfSuccessFull);
      const r = allStates.slice(indexOfSuccessFull);
      return l.concat([getSuccessfulExportProcessState(allStates.length)]).concat(r);
    }

    test.each([
      [
        "is the last in the export process state list",
        NumberOfNotSuccessfulExportProcessStates,
        getTestExportProcessStates(NumberOfNotSuccessfulExportProcessStates),
      ],
      ["is the first in the export process state list", 0, getTestExportProcessStates(0)],
      [
        "is in the middle of the export process state list",
        Math.round(NumberOfNotSuccessfulExportProcessStates / 2),
        getTestExportProcessStates(Math.round(NumberOfNotSuccessfulExportProcessStates / 2)),
      ],
      ["is the only one in the export process state list", 0, [getSuccessfulExportProcessState(0)]],
    ])(
      "should render the download button when the latest successful %s",
      (description, givenIndexOfSuccessFull, givenExportProcessStates) => {
        // GIVEN some model with the given export process states
        const givenModel = getOneFakeModel(1);
        givenModel.exportProcessState = givenExportProcessStates;

        // WHEN the component is rendered
        render(<ExportStateCellContent model={givenModel} />);

        // THEN expect no errors or warning to have occurred
        expect(console.error).not.toHaveBeenCalled();
        expect(console.warn).not.toHaveBeenCalled();

        // AND expect the download button to be rendered
        const actualElement = screen.getByTestId("mock-DownLoadButton-icon");
        expect(actualElement).toBeInTheDocument();

        // AND to match the snapshot
        expect(actualElement).toMatchSnapshot();

        // AND to have been called with the correct props
        expect(DownloadModelButton as jest.Mock).toHaveBeenCalledWith(
          {
            downloadUrl: givenModel.exportProcessState[givenIndexOfSuccessFull].downloadUrl,
          },
          {}
        );
      }
    );
  });

  describe.each(
    getAllNotSuccessfulExportProcessStates().map((exportProcessState, index) => {
      return [index, exportProcessState];
    })
  )(
    `should render the export process state icon when the last export was not successful`,
    (index, givenExportStatus) => {
      test(`${index} - ${givenExportStatus.status} - ${JSON.stringify(givenExportStatus.result)}`, () => {
        // GIVEN some model with no export process state
        const givenModel = getOneFakeModel(1);
        givenModel.exportProcessState = [givenExportStatus];

        // WHEN the component is rendered
        render(<ExportStateCellContent model={givenModel} />);

        // THEN expect no errors or warning to have occurred
        expect(console.error).not.toHaveBeenCalled();
        expect(console.warn).not.toHaveBeenCalled();

        // AND expect the download button to be rendered
        const actualElement = screen.getByTestId("mock-ExportProcessState-icon");
        expect(actualElement).toBeInTheDocument();

        // AND to match the snapshot
        expect(actualElement).toMatchSnapshot();

        // AND to have been called with the correct props
        expect(ExportProcessStateIcon as jest.Mock).toHaveBeenCalledWith(
          {
            exportProcessState: givenModel.exportProcessState[0],
          },
          {}
        );
      });
    }
  );
});
