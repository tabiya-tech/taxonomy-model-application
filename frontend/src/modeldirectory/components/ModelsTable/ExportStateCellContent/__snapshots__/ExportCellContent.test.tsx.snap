// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportStateCellContent should render the download button when the last export was successful should render the download button when the latest successful is in the middle of the export process state list 1`] = `
<div
  data-testid="mock-DownLoadButton-icon"
/>
`;

exports[`ExportStateCellContent should render the download button when the last export was successful should render the download button when the latest successful is the first in the export process state list 1`] = `
<div
  data-testid="mock-DownLoadButton-icon"
/>
`;

exports[`ExportStateCellContent should render the download button when the last export was successful should render the download button when the latest successful is the last in the export process state list 1`] = `
<div
  data-testid="mock-DownLoadButton-icon"
/>
`;

exports[`ExportStateCellContent should render the download button when the last export was successful should render the download button when the latest successful is the only one in the export process state list 1`] = `
<div
  data-testid="mock-DownLoadButton-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 0 - pending - {"errored":true,"exportErrors":true,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 1 - pending - {"errored":true,"exportErrors":true,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 2 - pending - {"errored":true,"exportErrors":false,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 3 - pending - {"errored":true,"exportErrors":false,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 4 - pending - {"errored":false,"exportErrors":true,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 5 - pending - {"errored":false,"exportErrors":true,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 6 - pending - {"errored":false,"exportErrors":false,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 7 - pending - {"errored":false,"exportErrors":false,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 8 - running - {"errored":true,"exportErrors":true,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 9 - running - {"errored":true,"exportErrors":true,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 10 - running - {"errored":true,"exportErrors":false,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 11 - running - {"errored":true,"exportErrors":false,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 12 - running - {"errored":false,"exportErrors":true,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 13 - running - {"errored":false,"exportErrors":true,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 14 - running - {"errored":false,"exportErrors":false,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 15 - running - {"errored":false,"exportErrors":false,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 16 - completed - {"errored":true,"exportErrors":true,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 17 - completed - {"errored":true,"exportErrors":true,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 18 - completed - {"errored":true,"exportErrors":false,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 19 - completed - {"errored":true,"exportErrors":false,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 20 - completed - {"errored":false,"exportErrors":true,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 21 - completed - {"errored":false,"exportErrors":true,"exportWarnings":false} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;

exports[`ExportStateCellContent should render the export process state icon when the last export was not successful 22 - completed - {"errored":false,"exportErrors":false,"exportWarnings":true} 1`] = `
<div
  data-testid="mock-ExportProcessState-icon"
/>
`;
